### DISCUSSION TITLE
Scan all image layers and all packages

### DISCUSSION THREAD

**Discussion by NikolayStad** [aquasecurity/trivy#4971] [Q&A] [answered]
Hello.
There are many tutorials on the net on how to fool security scanners.
For images based on known operating systems, it is enough to replace or clean up /var /lib /****** or other files.
How to protect yourself from this problem?
In trivy, scanning with the --list-all-pkgs flag does not help.
Syft with the collection of sbom for all layers partially closes the problem, but not all.

**Comment by knqyf263** [comment, marked as answer]
Trivy also collects data from all layers. But if the packages are removed in a single layer, it doesn't help anyway. First of all, Trivy is not designed to scan malicious images. It targets vulnerable images.

**Comment by NikolayStad** [comment]
Thank you!
Tried to scan trivy image as filesystem from inside. But trivy didn't find any OS packages either, with var /lib /dpkg removed.

Tell me, is it possible to somehow close the threat from malicious images using scanners? If you implement this functionality, then I'm ready to join the study.

**Reply by AnaisUrlichs** [reply]
There are other steps you can take to avoid using malicious container images; in terms of tools, there are other tools that focus on malware scanning in container images

**Comment by NikolayStad** [comment]
Thank you.
Yes, there are many virus scanners out there. But I have not seen a single scanner that looks for binary packages in all layers and packages not installed through the package manager

**Reply by AnaisUrlichs** [reply]
I think that would have to be a very opinionated tool to make automated statements based on missing information on "interesting file combinations"

**Comment by NikolayStad** [comment]
Yes you are right.
But there is a problem and cyber threats are growing.
This will incur additional resource overhead, but it makes sense when checking images when they are brought into the development loop, in order to be on the safe side.
The principle should be like that of an antivirus, detection of all binary files.

**Reply by knqyf263** [reply]
If you want to detect malicious images, you should verify signatures. Also, anti-virus is out of scope in Trivy OSS. Our commercial edition has the functionality.
